gems/aws-sdk-core/lib/aws-sdk-core/instance_profile_credentials.rb

```diff
@@ -97,260 +76,128 @@ def initialize(options = {})
     # @return [Integer] Number of times to retry when retrieving credentials
     #   from the instance metadata service. Defaults to 0 when resolving from
     #   the default credential chain ({Aws::CredentialProviderChain}).
-    attr_reader :retries
+    def retries
+      @client.retries
+    end
 
     private
 
-    def resolve_endpoint_mode(options)
-      value = options[:endpoint_mode]
-      value ||= ENV['AWS_EC2_METADATA_SERVICE_ENDPOINT_MODE']
-      value ||= Aws.shared_config.ec2_metadata_service_endpoint_mode(
-        profile: options[:profile]
-      )
-      value || 'IPv4'
+    def create_client(options)
+      endpoint_mode = resolve_endpoint_mode(options)
+      endpoint = resolve_endpoint(options)
+      options.merge!(endpoint_mode: endpoint_mode, endpoint: endpoint)
+      EC2Metadata.new(options)
     end
 
-    def resolve_endpoint(options, endpoint_mode)
-      value = options[:endpoint] || options[:ip_address]
-      value ||= ENV['AWS_EC2_METADATA_SERVICE_ENDPOINT']
-      value ||= Aws.shared_config.ec2_metadata_service_endpoint(
-        profile: options[:profile]
-      )
-
-      return value if value
+    def resolve_disable_ec2_metadata(options)
+      value =
+        ENV['AWS_EC2_METADATA_DISABLED'] ||
+        Aws.shared_config.disable_ec2_metadata(profile: options[:profile]) ||
+        'false'
+      Aws::Util.str_2_bool(value)
+    end
```

Comment:
I'm not sure why this existed here, but I think AWS_EC2_METADATA_DISABLED belongs in the credentials provider chain and not in this class. There's no use creating this class when it's disabled - I think the chain should check this, and skip making an instance of it instead.

Reply:
That's fair point. I think we would also want to handle cases where customers creates this instance directly - rather than relying the credential provider chain